src/jira/util/jwt.ts

```diff
@@ -94,25 +131,16 @@ export const hasValidJwt = (secret: string, baseUrl: string, req: Request, res:
 		return false;
 	}
 
-	// First check query string params
 	if (verifiedClaims.qsh) {
-		let expectedHash = jwt.createQueryStringHash(req, false, baseUrl);
-		let signatureHashVerified = verifiedClaims.qsh === expectedHash;
-		if (!signatureHashVerified) {
-			jwt.createCanonicalRequest(req, false, baseUrl); // eslint-disable-line
-
-			// If that didn't verify, it might be a post/put - check the request body too
-			expectedHash = jwt.createQueryStringHash(req, true, baseUrl);
-			signatureHashVerified = verifiedClaims.qsh === expectedHash;
-			if (!signatureHashVerified) {
-				jwt.createCanonicalRequest(req, true, baseUrl); // eslint-disable-line
-
-				// Send the error message for the first verification - it's 90% more likely to be the one we want.
-				sendError(res, 401, "Authentication failed: query hash does not match.");
-				return false;
-			}
+		if (tokenType === TokenType.context) {
+			//If we use context jsw tokens, their qsh will be constant
+			return verifiedClaims.qsh === "context-qsh"
+		} else {
+			//validate query string hash
+			return verifyQsh(verifiedClaims.qsh, baseUrl, req, res);
 		}
+	} else {
+		sendError(res, 401, "JWT tokens without qsh are not allowed");
+		return false
 	}
-
-	return true;
 };
```

Comment:
Surprisingly this file never had unit tests, so

- [ ] I'll add some tests for it